File: dashboards/backend/src/db/pmStore.ts
import { db } from "./index";
import { sql } from "kysely";
import type { ToolLifeTable, PreventiveMaintenanceTable } from "./types";

// ── Public-facing shapes (kept compatible with frontend) ───────────

export interface MaintenanceRecord {
  id: number;
  date: string;
  currentStroke: number;
  nextStroke: number;
  attachment: string | null;
}

export interface PMEntry {
  toolId: number;
  toolNo: string;
  toolLife: number;
  spm: number;
  pmStrokes: number;
  maintenanceHistory: MaintenanceRecord[];
  createdAt: string;
}

// ── Helpers ────────────────────────────────────────────────────────

function formatDate(d: string | Date): string {
  const date = typeof d === "string" ? new Date(d) : d;
  return date.toISOString();
}

async function resolveToolNumber(toolId: number): Promise<string | null> {
  const fromToolLife = await db
    .selectFrom("tool_life")
    .select("TL_tool_number as toolNo")
    .where("TL_tool_id", "=", toolId)
    .executeTakeFirst();

  if (fromToolLife?.toolNo) {
    return fromToolLife.toolNo;
  }

  const fromToolMaster = await db
    .selectFrom("components_tool")
    .select("CT_TOOLNO as toolNo")
    .where("CT_ID", "=", toolId)
    .executeTakeFirst();

  return fromToolMaster?.toolNo ?? null;
}

async function getToolStrokesByToolNo(toolNo: string): Promise<number> {
  const result = await sql<{ totalQty: number }>`
    SELECT COALESCE(MAX(comp.componentStrokes), 0) AS totalQty
    FROM (
      SELECT
        ct.CT_COMPID,
        SUM(pd.PD_PRODQTY / GREATEST(COALESCE(ct.CT_NO_OF_CAVITY, 1), 1)) AS componentStrokes
      FROM production_details pd
      INNER JOIN components_tool ct ON ct.CT_ID = pd.PD_TOOLID
      WHERE ct.CT_TOOLNO = ${toolNo}
      GROUP BY ct.CT_COMPID
    ) comp
  `.execute(db);

  return Number(result.rows[0]?.totalQty ?? 0);
}

// ── Public API ─────────────────────────────────────────────────────

/** Get all tool_life rows with their maintenance history attached. */
export async function getEntries(): Promise<PMEntry[]> {
  const tools = await db
    .selectFrom("tool_life")
    .selectAll()
    .execute();

  const pmRows = await db
    .selectFrom("preventive_maintenance")
    .selectAll()
    .orderBy("PM_id", "asc")
    .execute();

  const historyByToolId = new Map<number, MaintenanceRecord[]>();
  for (const pm of pmRows) {
    const list = historyByToolId.get(pm.PM_tool_id) ?? [];
    list.push({
      id: pm.PM_id,
      date: formatDate(pm.PM_date),
      currentStroke: Number(pm.PM_current_stroke),
      nextStroke: Number(pm.PM_next_stroke),
      attachment: pm.PM_maintenance_attachment,
    });
    historyByToolId.set(pm.PM_tool_id, list);
  }

  return tools.map((t) => {
    const history = historyByToolId.get(t.TL_tool_id) ?? [];

    return {
      toolId: t.TL_tool_id,
      toolNo: t.TL_tool_number,
      toolLife: Number(t.TL_life_span),
      spm: t.TL_spm,
      pmStrokes: Number(t.TL_preventive_maintenance_strokes),
      maintenanceHistory: history,
      createdAt: formatDate(t.TL_created_at),
    };
  });
}

/** Add a new tool_life row. Optionally creates an initial preventive_maintenance record. Throws on duplicate toolId. */
export async function addEntry(
  toolId: number,
  toolNo: string,
  toolLife: number,
  spm: number,
  pmStrokes: number,
  nextStroke?: number
): Promise<PMEntry> {
  // Check for existing row (PK will also enforce this but gives a nicer message)
  const existing = await db
    .selectFrom("tool_life")
    .select("TL_tool_number")
    .where("TL_tool_number", "=", toolNo)
    .executeTakeFirst();

  if (existing) {
    throw new Error(`Tool ${toolNo} (ID: ${toolId}) is already added`);
  }

  await db
    .insertInto("tool_life")
    .values({
      TL_tool_id: toolId,
      TL_tool_number: toolNo,
      TL_life_span: toolLife,
      TL_spm: spm,
      TL_preventive_maintenance_strokes: pmStrokes,
    })
    .execute();

  // If nextStroke is provided, create an initial preventive_maintenance record
  if (nextStroke !== undefined) {
    const currentStroke = await getToolStrokes(toolId);
    await db
      .insertInto("preventive_maintenance")
      .values({
        PM_tool_id: toolId,
        PM_tool_number: toolNo,
        PM_date: new Date().toISOString().slice(0, 19).replace("T", " "),
        PM_current_stroke: currentStroke,
        PM_next_stroke: nextStroke,
        PM_maintenance_attachment: null,
      })
      .execute();
  }

  const entries = await getEntries();
  return entries.find((e) => e.toolId === toolId)!;
}

/** Update tool_life fields for a given toolId. */
export async function updateEntry(
  toolId: number,
  updates: { toolLife?: number; pmStrokes?: number; spm?: number }
): Promise<PMEntry> {
  const row = await db
    .selectFrom("tool_life")
    .selectAll()
    .where("TL_tool_id", "=", toolId)
    .executeTakeFirst();

  if (!row) {
    throw new Error(`Tool ID ${toolId} not found`);
  }

  const setValues: Record<string, unknown> = {};
  if (updates.toolLife !== undefined) setValues.TL_life_span = updates.toolLife;
  if (updates.pmStrokes !== undefined) setValues.TL_preventive_maintenance_strokes = updates.pmStrokes;
  if (updates.spm !== undefined) setValues.TL_spm = updates.spm;

  if (Object.keys(setValues).length > 0) {
    await db
      .updateTable("tool_life")
      .set(setValues)
      .where("TL_tool_id", "=", toolId)
      .execute();
  }

  // Fetch updated record with history
  const entries = await getEntries();
  const entry = entries.find((e) => e.toolId === toolId);
  return entry!;
}

/** Get total strokes for any tool from production_details (does NOT require tool_life entry). */
export async function getToolStrokes(toolId: number): Promise<number> {
  const toolNo = await resolveToolNumber(toolId);
  if (!toolNo) {
    return 0;
  }

  return getToolStrokesByToolNo(toolNo);
}

/** Get current total strokes and suggested next PM stroke for a given toolId. */
export async function getStrokeInfo(
  toolId: number
): Promise<{ currentStroke: number; suggestedNextStroke: number }> {
  const row = await db
    .selectFrom("tool_life")
    .selectAll()
    .where("TL_tool_id", "=", toolId)
    .executeTakeFirst();

  if (!row) {
    throw new Error(`Tool ID ${toolId} not found`);
  }

  const currentStroke = await getToolStrokes(toolId);
  const suggestedNextStroke = currentStroke + Number(row.TL_preventive_maintenance_strokes);

  return { currentStroke, suggestedNextStroke };
}

/** Record a maintenance event for a given toolId. Computes current stroke from production_details. */
export async function confirmMaintenance(
  toolId: number,
  nextStroke: number,
  attachment?: string
): Promise<PMEntry> {
  const row = await db
    .selectFrom("tool_life")
    .selectAll()
    .where("TL_tool_id", "=", toolId)
    .executeTakeFirst();

  if (!row) {
    throw new Error(`Tool ID ${toolId} not found`);
  }

  // Compute current total strokes from production_details
  const currentStroke = await getToolStrokes(toolId);

  await db
    .insertInto("preventive_maintenance")
    .values({
      PM_tool_id: toolId,
      PM_tool_number: row.TL_tool_number,
      PM_date: new Date().toISOString().slice(0, 19).replace("T", " "),
      PM_current_stroke: currentStroke,
      PM_next_stroke: nextStroke,
      PM_maintenance_attachment: attachment ?? null,
    })
    .execute();

  const entries = await getEntries();
  return entries.find((e) => e.toolId === toolId)!;
}

/** Delete a tool_life row (and cascade will handle PM rows via RESTRICT — must delete PM rows first). */
export async function deleteEntry(toolId: number): Promise<void> {
  const row = await db
    .selectFrom("tool_life")
    .select("TL_tool_id")
    .where("TL_tool_id", "=", toolId)
    .executeTakeFirst();

  if (!row) {
    throw new Error(`Tool ID ${toolId} not found`);
  }

  // Delete related preventive_maintenance rows first (FK is RESTRICT)
  await db
    .deleteFrom("preventive_maintenance")
    .where("PM_tool_id", "=", toolId)
    .execute();

  await db
    .deleteFrom("tool_life")
    .where("TL_tool_id", "=", toolId)
    .execute();
}

File: backend/pm_store.py
"""Preventive Maintenance data-access layer.

Port of dashboards/backend/src/db/pmStore.ts — uses the existing db.py
helpers (fetch_all, fetch_one, execute) with raw SQL.
"""

from __future__ import annotations

import os
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional
from urllib.parse import quote, unquote

from .db import execute, fetch_all, fetch_one


# ── Helpers ─────────────────────────────────────────────────────────────

def _format_date(d: Any) -> str:
    if isinstance(d, datetime):
        return d.isoformat()
    if isinstance(d, str):
        try:
            return datetime.fromisoformat(d).isoformat()
        except Exception:
            return d
    return str(d)


def _resolve_tool_number(tool_id: int) -> Optional[str]:
    row = fetch_one(
        "SELECT TL_tool_number AS toolNo FROM tool_life WHERE TL_tool_id = %s",
        (tool_id,),
    )
    if row and row.get("toolNo"):
        return row["toolNo"]

    row = fetch_one(
        "SELECT CT_TOOLNO AS toolNo FROM components_tool WHERE CT_ID = %s",
        (tool_id,),
    )
    return row["toolNo"] if row and row.get("toolNo") else None


def _get_tool_strokes_by_tool_no(tool_no: str) -> int:
    row = fetch_one(
        """
        SELECT COALESCE(MAX(comp.componentStrokes), 0) AS totalQty
        FROM (
            SELECT
                ct.CT_COMPID,
                SUM(pd.PD_PRODQTY / GREATEST(COALESCE(ct.CT_NO_OF_CAVITY, 1), 1)) AS componentStrokes
            FROM production_details pd
            INNER JOIN components_tool ct ON ct.CT_ID = pd.PD_TOOLID
            WHERE ct.CT_TOOLNO = %s
            GROUP BY ct.CT_COMPID
        ) comp
        """,
        (tool_no,),
    )
    return int(row["totalQty"]) if row else 0


def _safe_pm_attachment_relpath(raw: str) -> Optional[str]:
    """Build a safe relative path (forward slashes) under pm-attachments; None if invalid."""
    s = unquote(raw.strip()).replace("\\", "/")
    if not s:
        return None
    s = s.lstrip("/")
    low = s.lower()
    if low.startswith("pm-attachments/"):
        s = s[len("pm-attachments/") :]
    parts: List[str] = []
    for p in s.split("/"):
        if not p or p == ".":
            continue
        if p == "..":
            return None
        parts.append(p)
    return "/".join(parts) if parts else None


def _normalize_attachment_path(value: Optional[str]) -> Optional[str]:
    """Return API-served PM attachment URL for legacy/raw DB values.

    Preserves nested folders (e.g. ``S1/PT041/01A/file.pdf``) instead of basename-only,
    so URLs match files stored under ``pm-attachments/<subdirs>/``.
    """
    if not value:
        return None

    attachment = str(value).strip()
    if not attachment:
        return None

    if attachment.startswith(("http://", "https://")):
        return attachment

    if attachment.startswith("/api/pm/attachment/"):
        inner = attachment[len("/api/pm/attachment/") :]
        safe = _safe_pm_attachment_relpath(inner)
        if not safe:
            return attachment
        return f"/api/pm/attachment/{quote(safe, safe='/')}"

    safe = _safe_pm_attachment_relpath(attachment)
    if not safe:
        return None
    return f"/api/pm/attachment/{quote(safe, safe='/')}"


# ── Public API ──────────────────────────────────────────────────────────

def get_entries() -> List[Dict[str, Any]]:
    """Get all tool_life rows with their maintenance history attached."""
    tools = fetch_all("SELECT * FROM tool_life")
    pm_rows = fetch_all(
        "SELECT * FROM preventive_maintenance ORDER BY PM_id ASC"
    )

    history_by_tool_id: Dict[int, List[Dict[str, Any]]] = {}
    for pm in pm_rows:
        tid = pm["PM_tool_id"]
        rec = {
            "id": pm["PM_id"],
            "date": _format_date(pm["PM_date"]),
            "currentStroke": int(pm["PM_current_stroke"]),
            "nextStroke": int(pm["PM_next_stroke"]),
            "attachment": _normalize_attachment_path(pm.get("PM_maintenance_attachment")),
        }
        history_by_tool_id.setdefault(tid, []).append(rec)

    result = []
    for t in tools:
        tid = t["TL_tool_id"]
        history = history_by_tool_id.get(tid, [])
        result.append({
            "toolId": tid,
            "toolNo": t["TL_tool_number"],
            "toolLife": int(t["TL_life_span"]),
            "spm": int(t["TL_spm"]),
            "pmStrokes": int(t["TL_preventive_maintenance_strokes"]),
            "maintenanceHistory": history,
            "createdAt": _format_date(t["TL_created_at"]),
        })

    return result


def add_entry(
    tool_id: int,
    tool_no: str,
    tool_life: int,
    spm: int,
    pm_strokes: int,
    next_stroke: Optional[int] = None,
) -> Dict[str, Any]:
    """Add a new tool_life row. Optionally creates an initial PM record."""
    existing = fetch_one(
        "SELECT TL_tool_number FROM tool_life WHERE TL_tool_number = %s",
        (tool_no,),
    )
    if existing:
        raise ValueError(f"Tool {tool_no} (ID: {tool_id}) is already added")

    execute(
        """
        INSERT INTO tool_life
            (TL_tool_id, TL_tool_number, TL_life_span, TL_spm, TL_preventive_maintenance_strokes)
        VALUES (%s, %s, %s, %s, %s)
        """,
        (tool_id, tool_no, tool_life, spm, pm_strokes),
    )

    if next_stroke is not None:
        current_stroke = get_tool_strokes(tool_id)
        now_str = datetime.now(timezone.utc).strftime("%Y-%m-%d %H:%M:%S")
        execute(
            """
            INSERT INTO preventive_maintenance
                (PM_tool_id, PM_tool_number, PM_date, PM_current_stroke, PM_next_stroke, PM_maintenance_attachment)
            VALUES (%s, %s, %s, %s, %s, %s)
            """,
            (tool_id, tool_no, now_str, current_stroke, next_stroke, None),
        )

    entries = get_entries()
    return next((e for e in entries if e["toolId"] == tool_id), entries[0])


def update_entry(
    tool_id: int,
    updates: Dict[str, Any],
) -> Dict[str, Any]:
    """Update tool_life fields for a given toolId."""
    row = fetch_one(
        "SELECT * FROM tool_life WHERE TL_tool_id = %s",
        (tool_id,),
    )
    if not row:
        raise ValueError(f"Tool ID {tool_id} not found")

    set_parts = []
    params: list = []
    if "toolLife" in updates and updates["toolLife"] is not None:
        set_parts.append("TL_life_span = %s")
        params.append(updates["toolLife"])
    if "pmStrokes" in updates and updates["pmStrokes"] is not None:
        set_parts.append("TL_preventive_maintenance_strokes = %s")
        params.append(updates["pmStrokes"])
    if "spm" in updates and updates["spm"] is not None:
        set_parts.append("TL_spm = %s")
        params.append(updates["spm"])

    if set_parts:
        params.append(tool_id)
        execute(
            f"UPDATE tool_life SET {', '.join(set_parts)} WHERE TL_tool_id = %s",
            tuple(params),
        )

    entries = get_entries()
    entry = next((e for e in entries if e["toolId"] == tool_id), None)
    if not entry:
        raise ValueError(f"Tool ID {tool_id} not found after update")
    return entry


def get_tool_strokes(tool_id: int) -> int:
    """Get total strokes for any tool from production_details."""
    tool_no = _resolve_tool_number(tool_id)
    if not tool_no:
        return 0
    return _get_tool_strokes_by_tool_no(tool_no)


def get_stroke_info(tool_id: int) -> Dict[str, int]:
    """Get current total strokes and suggested next PM stroke."""
    row = fetch_one(
        "SELECT * FROM tool_life WHERE TL_tool_id = %s",
        (tool_id,),
    )
    if not row:
        raise ValueError(f"Tool ID {tool_id} not found")

    current_stroke = get_tool_strokes(tool_id)
    suggested = current_stroke + int(row["TL_preventive_maintenance_strokes"])
    return {"currentStroke": current_stroke, "suggestedNextStroke": suggested}


def confirm_maintenance(
    tool_id: int,
    next_stroke: int,
    attachment: Optional[str] = None,
) -> Dict[str, Any]:
    """Record a maintenance event."""
    row = fetch_one(
        "SELECT * FROM tool_life WHERE TL_tool_id = %s",
        (tool_id,),
    )
    if not row:
        raise ValueError(f"Tool ID {tool_id} not found")

    current_stroke = get_tool_strokes(tool_id)
    now_str = datetime.now(timezone.utc).strftime("%Y-%m-%d %H:%M:%S")

    execute(
        """
        INSERT INTO preventive_maintenance
            (PM_tool_id, PM_tool_number, PM_date, PM_current_stroke, PM_next_stroke, PM_maintenance_attachment)
        VALUES (%s, %s, %s, %s, %s, %s)
        """,
        (tool_id, row["TL_tool_number"], now_str, current_stroke, next_stroke, attachment),
    )

    entries = get_entries()
    return next((e for e in entries if e["toolId"] == tool_id), entries[0])


def delete_entry(tool_id: int) -> None:
    """Delete a tool_life row and its maintenance records."""
    row = fetch_one(
        "SELECT TL_tool_id FROM tool_life WHERE TL_tool_id = %s",
        (tool_id,),
    )
    if not row:
        raise ValueError(f"Tool ID {tool_id} not found")

    execute(
        "DELETE FROM preventive_maintenance WHERE PM_tool_id = %s",
        (tool_id,),
    )
    execute(
        "DELETE FROM tool_life WHERE TL_tool_id = %s",
        (tool_id,),
    )
